executor/join.go

```diff
@@ -512,26 +532,31 @@ func (e *HashJoinExec) fetchInnerAndBuildHashTable(ctx context.Context) {
 		}
 		close(e.innerFinished)
 	}()
-
-	if err := e.fetchInnerRows(ctx); err != nil {
-		e.innerFinished <- errors.Trace(err)
-		return
-	}
+	// innerResultCh transfer inner result chunk from inner fetch to build hash table.
+	innerResultCh := make(chan *chunk.Chunk, e.concurrency)
+	doneCh := make(chan struct{})
+	go e.fetchInnerRows(ctx, innerResultCh, doneCh)
```

Comment:
how about don't always fork new groutine in here and keep fetchInnerRows logic in "primary" goroutine, and base on fetch result to choose whether or not to fork new groutine to buildHashTableForList.(e.g. Next 2 times but still has data?)

IMHO, The idea is that maybe common small innser case it's too heavy to fork goroutine but some case fork is worth, and fetch with single goutine is required, so make choosen in buildTable?

Reply:
@crazycs520 We can bench the scenario that the inner child for the hash join operator is a simple table reader and the number of output records of that child is very small, for example, 1.

Reply:
@lysu @zz-jason here is a benchmark:
`select t1.* from t1 inner join tid1 where t1.id=tid1.id and tid1.id < 1;`
table `t1` and `tid1` both have 10 rows. and the join result is 1 row;
master:
```sql
query 0 takes 0.0160129070282s
query 0 takes 0.0165090560913s
query 0 takes 0.0154628753662s
query 0 takes 0.0160419940948s
query 0 takes 0.0180399417877s
avgTime 0.0164133548737s
```
this branch
```
query 0 takes 0.02263879776s
query 0 takes 0.0164089202881s
query 0 takes 0.0150151252747s
query 0 takes 0.0146901607513s
query 0 takes 0.0151329040527s
avgTime 0.0167771816254s
```